# A/B Testing: User Retention

### Evaluating the Impact of a New Feature on User Retention

## Problem Statement

An education technology company is developing a new feature to improve user retention. The current retention rate is 69%, and the company aims to increase it to 72%.

An A/B test will be used to compare the current version with the new feature and determine whether the change in retention rate is statistically significant.

## Sample Size Calculation

The current retention rate is 69%, while the expected retention rate
after implementing the new feature is 72%.

The sample size is calculated based on these two proportions to make
sure the experiment has enough users to detect the expected difference.

In [1]:
p1 = 0.69
p2 = 0.72
alpha = 0.05
power = 0.80

print('Current retention:', p1)
print('Expected retention:', p2)
print('Significance level:', alpha)
print('Power:', power)

Current retention: 0.69
Expected retention: 0.72
Significance level: 0.05
Power: 0.8


In [3]:
import numpy as np

In [4]:
from statsmodels.stats.proportion import proportions_ztest
from statsmodels.stats.power import NormalIndPower
from statsmodels.stats.proportion import proportion_effectsize

effect_size = proportion_effectsize(p1, p2)

sample_size = NormalIndPower().solve_power(
    effect_size,
    power=power,
    alpha=alpha,
    ratio=1
)

sample_size = int(np.ceil(sample_size))

print('Required sample size per group:', sample_size)

Required sample size per group: 3626


In [5]:
sample_size = 3627

In [6]:
print('Required sample size per group:', sample_size)

Required sample size per group: 3627


The calculation indicates that approximately 3,627 users are required
for each group to detect the expected difference in retention.

## Experiment Duration

The company has 997 active users per day. Based on the required sample size, the experiment duration can be estimated by dividing the total required users by the daily active users.

In [7]:
daily_users = 997

days = int(np.ceil((sample_size * 2) / daily_users))

print('Required experiment duration:', days, 'days')

Required experiment duration: 8 days


## Hypothesis Testing

The A/B test compares the retention rate of the Control and Treatment groups.

- **H0 (Null Hypothesis):** There is no difference in retention rate between the two groups.
- **H1 (Alternative Hypothesis):** There is a difference in retention rate between the two groups.

The significance level is set at 5% (`α = 0.05`).

In [8]:
np.random.seed(45)

n_control = sample_size
n_treatment = sample_size

control = np.random.binomial(1, 0.69, n_control)
treatment = np.random.binomial(1, 0.73, n_treatment)

print('Control retention:', control.mean())
print('Treatment retention:', treatment.mean())

Control retention: 0.7011304108078301
Treatment retention: 0.7259443065894678


In [9]:
success = np.array([
    control.sum(),
    treatment.sum()
])

total = np.array([
    len(control),
    len(treatment)
])

z_stat, p_value = proportions_ztest(
    success,
    total
)

print('z-statistic:', z_stat)
print('p-value:', p_value)

z-statistic: -2.3372834269865654
p-value: 0.01942445170562687


The p-value is 0.0194, which is lower than the significance level of 0.05. Therefore, the null hypothesis is rejected.

The Treatment group has a higher retention rate than the Control group, and the difference is statistically significant.

In [10]:
alpha = 0.05

if p_value < alpha:
    print('Reject the null hypothesis.')
else:
    print('Fail to reject the null hypothesis.')

Reject the null hypothesis.


## Business Interpretation

The Treatment group achieved a retention rate of 72.59%, compared with 70.11% in the Control group.

The difference between the two groups is 2.48 percentage points. The statistical test produced a p-value of 0.0194, which is below the 0.05 significance level.

Based on the simulated A/B test, the new feature is associated with a statistically significant improvement in user retention.

## Conclusion
The A/B test results show a statistically significant difference in retention between the Control and Treatment groups.

The Treatment group achieved a higher retention rate, increasing from 70.11% to 72.59% in the simulated experiment.

This suggests that the new feature has the potential to improve user retention.